File: libdrqueue/computer_info.c
//
// Copyright (C) 2001,2002,2003,2004,2005,2006,2007 Jorge Daza Garcia-Blanes
//
// This file is part of DrQueue
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
// USA
//

#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <sys/types.h>
#include <signal.h>
#include <stdint.h>
#include <logger.h>

#ifdef __IRIX
#include <sys/sysmp.h>
#endif

#ifdef __FREEBSD
#include <sys/sysctl.h>
#endif

#include "pointer.h"
#include "constants.h"
#include "computer_info.h"

#if defined (__LINUX)
#include "computer_info.linux.h"
#elif defined (__IRIX)
#include "computer_info.irix.h"
#elif defined (__OSX)
#include "computer_info.osx.h"
#elif defined (__FREEBSD)
#include "computer_info.freebsd.h"
#elif defined (__CYGWIN)
#include "computer_info.cygwin.h"
#else
#error You need to define the OS, or OS defined not supported
#endif

uint32_t
get_speedindex (struct computer_hwinfo *hwinfo) {
  uint32_t speedindex;

  speedindex = (hwinfo->proctype + 1) * hwinfo->procspeed;

  return speedindex;
}


void report_hwinfo (struct computer_hwinfo *hwinfo) {
  printf ("HWINFO Report\n");
  printf ("Name:\t\t\t%s\n",hwinfo->name);
  printf ("Architecture:\t\t%s\n",archstring((t_arch)hwinfo->arch));
  printf ("OS:\t\t\t%s\n",osstring((t_os)hwinfo->os));
  printf ("Processor type:\t\t%s\n",proctypestring((t_proctype)hwinfo->proctype));
  printf ("64bit-32bit cpu:\t%s\n",bitsstring(hwinfo->nnbits));
  printf ("Processor speed:\t%i MHz\n",hwinfo->procspeed);
  printf ("Number of processors:\t%i\n",hwinfo->ncpus);
  //printf ("Speed index:\t\t%i\n",hwinfo->speedindex);
  printf ("Memory:\t\t\t%i Mbytes\n",hwinfo->memory);
}

char *osstring (t_os os) {
  char *msg;

  switch (os) {
  case OS_UNKNOWN:
    msg = "UNKNOWN";
    break;
  case OS_IRIX:
    msg = "Irix";
    break;
  case OS_LINUX:
    msg = "Linux";
    break;
  case OS_WINDOWS:
  case OS_CYGWIN:
    msg = "Windows";
    break;
  case OS_OSX:
    msg = "Mac OSX";
    break;
  case OS_FREEBSD:
    msg = "FreeBSD";
    break;
  default:
    msg = "DEFAULT (ERROR)";
    fprintf (stderr,"os == DEFAULT\n");
  }

  return msg;
}


uint8_t
computer_info_nnbits () {
  // This function will try to guess if the computer is a 32 or 64 bit
  // one. For that we'll use the size of a pointer to void
  int bytes=sizeof(void*);
  if (bytes == 4) {
    // This seems to be a 32bit cpu (or a 64bit one emulating 32bit ?)
    return 32;
  } else if (bytes == 8) {
    // Well, a 64bit one is my best guess
    return 64;
  }

  // In other situations:
  return 0;
}

char *
bitsstring (uint8_t nnbits) {
  char *msg;

  switch (nnbits) {
  case 0:
    msg = "Unknown";
    break;
  case 32:
    msg = "32bit";
    break;
  case 64:
    msg = "64bit";
    break;
  default:
    msg = "ERROR: value not listed";
  }

  return msg;
}

char *archstring (t_arch arch) {
  char *msg;

  switch (arch) {
  case ARCH_UNKNOWN:
    msg = "UNKNOWN";
    break;
  case ARCH_INTEL:
    msg = "Intel";
    break;
  case ARCH_MIPS:
    msg = "Mips";
    break;
  case ARCH_PPC:
    msg = "PowerPC";
    break;
  case ARCH_SPARC:
    msg = "Sparc";
    break;
  case ARCH_POWER:
    msg = "Power";
    break;
  default:
    msg = "DEFAULT (ERROR)";
    fprintf (stderr,"arch == DEFAULT\n");
  }

  return msg;
}

char *proctypestring (t_proctype proctype) {
  char *msg;

  switch (proctype) {
  case PROCTYPE_UNKNOWN:
    msg = "UNKNOWN";
    break;
  case PROCTYPE_INTEL_UNKNOWN:
    msg = "Intel (not listed)";
    break;
  case PROCTYPE_PENTIUM:
    msg = "Pentium";
    break;
  case PROCTYPE_PENTIUMII:
    msg = "Pentium II";
    break;
  case PROCTYPE_PENTIUMIII:
    msg = "Pentium III";
    break;
  case PROCTYPE_PENTIUM4:
    msg = "Pentium 4";
    break;
  case PROCTYPE_INTELXEON:
    msg = "Xeon";
    break;
  case PROCTYPE_INTELIA64:
    msg = "IA-64";
    break;
  case PROCTYPE_PENTIUMM:
    msg = "Pentium M";
    break;
  case PROCTYPE_ATHLON:
    msg = "Athlon";
    break;
  case PROCTYPE_OPTERON:
    msg = "Opteron";
    break;
  case PROCTYPE_MIPSR5000:
    msg = "R5000";
    break;
  case PROCTYPE_MIPSR10000:
    msg = "R10000";
    break;
  case PROCTYPE_MIPSR12000:
    msg = "R12000";
    break;
  case PROCTYPE_PPC:
    msg = "PPC";
    break;
  case PROCTYPE_ULTRASPARC:
    msg = "Ultrasparc";
    break;
  case PROCTYPE_CELLBE:
    msg = "Cell Broadband Engine";
    break;
  default:
    msg = "DEFAULT (ERROR)";
    fprintf (stderr,"proctype == DEFAULT\n");
  }

  return msg;
}

File: libdrqueue_python/computer_info.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''type info here'''
#
# Copyright (C) 2010 Andreas Schroeder
#
# This file is part of DrQueue
#
# DrQueue is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# DrQueue is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
# USA
#

# might use: constants


#typedef enum {
#  ARCH_UNKNOWN,
#  ARCH_INTEL,
#  ARCH_MIPS,
#  ARCH_PPC,
#  ARCH_SPARC,
#  ARCH_POWER
#} t_arch;
#
#typedef enum {
#  OS_UNKNOWN,
#  OS_IRIX,
#  OS_LINUX,
#  OS_CYGWIN,
#  OS_WINDOWS,
#  OS_OSX,
#  OS_FREEBSD
#} t_os;
#
#typedef enum {
#  PROCTYPE_UNKNOWN,
#  PROCTYPE_INTEL_UNKNOWN,
#  PROCTYPE_PENTIUM,
#  PROCTYPE_PENTIUMII,
#  PROCTYPE_PENTIUMIII,
#  PROCTYPE_PENTIUM4,
#  PROCTYPE_INTELXEON,
#  PROCTYPE_INTELCOREDUO,
#  PROCTYPE_INTELCORE2DUO,
#  PROCTYPE_INTELIA64,
#  PROCTYPE_PENTIUMM,
#  PROCTYPE_ATHLON,
#  PROCTYPE_OPTERON,
#  PROCTYPE_MIPSR5000,
#  PROCTYPE_MIPSR10000,
#  PROCTYPE_MIPSR12000,
#  PROCTYPE_PPC,
#  PROCTYPE_INTEL_CORE2,
#  PROCTYPE_ULTRASPARC,
#  PROCTYPE_CELLBE
#} t_proctype;
#
#struct computer_hwinfo {
#  char     name[MAXNAMELEN];   /* Name of the computer */
#  char     address[MAXNAMELEN]; /* cached IP address */ 
#  uint32_t id;                 /* Identification number */
#  uint8_t  arch;               /* type of architecture */
#  uint8_t  os;                 /* type of operating system */
#  uint8_t  proctype;           /* type of processors */
#  uint32_t procspeed;          /* speed of the processors */
#  uint16_t ncpus;              /* number of processors that the computer has */
#  uint32_t speedindex;         /* global speed index for making comparisons between different computers */
#  uint32_t memory;             /* Memory in Mbytes */
#  uint8_t  nnbits;             // 64/32 or 0 (Unknown) cpu
#};


class info:
  """Computer info objects"""

  def __init__(self):
    self.name = ""
    self.address = ""
    self.id = 0
    self.arch = 0
    self.os = 0
    self.proctype = 0
    self.procspeed = 0
    self.ncpus = 0
    self.speedindex = 0
    self.memory = 0
    self.nbits = 0

  # FIXME: will be accessed diretly by computer class
  #void get_hwinfo (struct computer_hwinfo *hwinfo);
  def get_hwinfo(self, hwinfo):
    return
  
  #t_proctype get_proctype (void);
  def get_proctype(self):
    return self.proctype
  
  #uint32_t get_procspeed (void);
  def get_procspeed(self):
    return self.procspeed
  
  #uint16_t get_numproc (void);
  def get_numproc(self):
    return self.numproc
  
  #uint32_t get_speedindex (struct computer_hwinfo *hwinfo);
  def get_speedindex(self, hwinfo):
    return self.speedindex
  
  #uint32_t get_memory (void);
  def get_memory(self):
    return self.memory
  
  #uint8_t computer_info_nnbits (void);
  def computer_info_nnbits(self):
    return self.nbits
  
  #void report_hwinfo (struct computer_hwinfo *hwinfo);
  def report_hwinfo(self, hwinfo):
    hwinfo = self
    return hwinfo
  
  #char *bitsstring (uint8_t nnbits);
  def bitsstring(nnbits):
    return ""
  
  #char *osstring (t_os os);
  def osstring(os):
    return ""
  
  #char *archstring (t_arch arch);
  def archstring(arch):
    return ""
  
  #char *proctypestring (t_proctype proctype);
  def proctypestring(proctype):
    return ""
  
  
